### ISSUE TITLE
REST API Inspection

### ISSUE THREAD

**Issue by dmlb2000** [EMSL-MSC/pacifica-metadata#35]
Users of the API should be able to inspect the url hierarchy from `/` and determine what valid URLs are through simple http gets.

**Comment by kfox1111** [comment]
Consider swagger support too. its gaining popularity as an easy way to look at the apis:
http://swagger.io/getting-started/

**Comment by dmlb2000** [comment]
swagger is interesting but it looks like it would require a complete rewrite of the rest code to lift it out of CherryPy and use Swagger... 

This seems like too much of a lift right now, I am hoping for something simpler...

**Comment by kfox1111** [comment]
I was under the (possibly wrong) impression you could just drop a document of the right format on the web server and swagger tools could pick it up. That its more of a document format for describing the rest calls supported then an engine that you use to write your rest api. Maybe I'm mistaken?

**Comment by dmlb2000** [comment]
The documentation I was reading wasn't very clear on the dilatation between what's swagger and what's other python library stuff...

The documentation for this sort of thing kinda reminds me of CORBA though using the REST and HTTP as transports instead of binary socket format... and there's two parts to a system like that, what the client is expected to do using a Swagger endpoint and how a server is supposed to implement a Swagger endpoint. The documentation wasn't clear on which side I was looking at...

Google `swagger example python server` and what you get is the `tools and integration` page... something is wrong with their searching...

Their petstore example was a full fledged rest service api thing. I couldn't find the documentation that told me what the parts where...

**Comment by kfox1111** [comment]
Yeah. I think they provide a bunch of tools to let you automatically generate swagger if you use them. otherwise, I think you have to craft the doc by hand, ala: http://swagger.io/specification/

But if your documenting by hand anyway, it might be useful to use their formatting, as some of their ui tools, such as http://swagger.io/swagger-ui/ are quite nice. I've used the swagger-ui against another project that was java based and was able to poke at its rest api pretty easily.

**Comment by dmlb2000** [comment]
There's only one reference to CherryPy and Swagger on the internets...

https://groups.google.com/forum/#!topic/cherrypy-users/095YSYOM_dI

So I think I'm going to wait until CherryPy has some sort of inspection logic built into it so that it can implement the Swagger spec based on a call to the CherryPy service somehow... 

https://github.com/cherrypy/cherrypy/issues/1507

I submitted an issue we'll see what they have to say about it.

**Comment by dmlb2000** [comment]
Just FYI I've started work on integration between CherryPy and Swagger at https://github.com/dmlb2000/cherrypy-swagger
